lib/libc/arcmwdt/libc-hooks.c

```diff
@@ -71,6 +71,13 @@ int _write(int fd, const char *buf, unsigned int nbytes)
 }
 #endif
 
+#ifndef CONFIG_POSIX_DEVICE_IO
+__weak int fileno(FILE *file)
+{
+	return _fileno(file);
+}
+#endif
```

Comment:
Does something like this work? I guess not many people have access to the arc mwdt toolchain to test 

```suggestion
__weak int _fileno(FILE *file)
{
	return zvfs_fileno(file);
}
```

Reply:
Well, we have a `_fileno()` function in our ARC toolchain, but not a `fileno()` function (it's a macro). So my idea was:
- Undefine the macro in any case because it prevents doing anything
- If CONFIG_POSIX_DEVICE_IO is defined, just use `fileno()` from the POSIX subsystem
- Otherwise define `fileno()` as a wrapper around our `_fileno()` to avoid user breakage
I can't see how your proposal solves the issue of missing `fileno()`. But if you have any suggestions regarding the situation I've described above, they will be very welcome.

Reply:
`fileno()` shouldn't be missing if `CONFIG_POSIX_DEVICE_IO` is under.

Reply:
Okay, so there are two separate questions: how to fix the issue of missing `fileno()` in the MWDT toolchain and whether we want to tie `_fileno()` to `zvfs_fileno()` unconditionally.
As to the first one, I still think what I did makes sense and I see no other obvious way to fix the issue.
But as to the second one, I'd like to hear your suggestions. Should we do this unconditionally or it should depend on something enabled? CONFIG_ZVFS? CONFIG_FDTABLE? Anything else?

Reply:
And, to be honest, I don't understand at all how the `fileno()` and `fdopen()` functions provided by ZVFS could ever be plugged into e.g. `fwrite()` from any embedded libc, either MWDT or Picolibc, because of different interpretations of what a `FILE` is. It might require changing libc integration layers completely.

Reply:
> But as to the second one, I'd like to hear your suggestions. Should we do this unconditionally or it should depend on something enabled? CONFIG_ZVFS? CONFIG_FDTABLE? Anything else?


Good catch, an `ifdef CONFIG_FDTABLE` would probably work

Reply:
Well, I've checked the source and found that the MWDT library doesn't use `_fileno()` internally, so there is no way to replace it with the ZVFS one. I'd really like to see someday a nice clean stdio sitting on top of the POSIX layer, but this is not the day yet. As for now, I propose to keep the change as is.

Reply:
> Does something like this work? I guess not many people have access to the arc mwdt toolchain to test

This is definitely wrong -- the FILE here is from libc, not the (invalid) FILE from zvfs.

Reply:
> I'd really like to see someday a nice clean stdio sitting on top of the POSIX layer, but this is not the day yet. 

That would be possible if Zephyr decided to become a POSIX OS.

Feel free to request that as an enhancement. I would support it.

Technically, POSIX is already mandatory for some types and Zephyr supports much of what is required already.

Reply:
> This is definitely wrong -- the FILE here is from libc, not the (invalid) FILE from zvfs.

If the libc `FILE*` contained a `void*` representing the Zephyr `fd_entry*`, then that would work.